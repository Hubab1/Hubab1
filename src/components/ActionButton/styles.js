import {
    css
} from 'emotion';
import styled from '@emotion/styled';

export const root = css`
    border-radius: 21.5px !important;
    height: 45px;
`;

export const label = css`
    text-transform: none;
    font-size: 16px;
`;

export const mainDisabled = css`
    background-color: #D5D8DE !important;
    color: #ffffff !important;
    box-shadow: 0px 1px 5px 0px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14), 0px 3px 1px -2px rgba(0,0,0,0.12) !important;
`;

export const ButtonContainer = styled.div(props => ({
    marginTop: props.marginTop || 0,
    marginBottom: props.marginBottom || 0,
    textDecoration: 'none',
    display: 'block',
}));
