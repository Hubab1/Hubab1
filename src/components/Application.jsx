import React from 'react';
import { Route, Switch } from 'react-router-dom';
import { buildRoute, getInitialPage } from 'utils/routeNavigation';
import { Routes } from 'constants.js';
import ProfileContainer from 'components/profile/ProfileContainer';
import TermsPage from 'components/TermsPage';
import AppContext from 'contexts/AppContext';
export default class Application extends React.Component {
    static contextType = AppContext;
    componentDidMount () {
        const { history, match } = this.props;
        const initialPage = getInitialPage(this.context.communityId, this.props.profile);
        history.replace(initialPage);
        history.replace('terms')
    }
    
    render () {
        const { match } = this.props;
        return (
            <Switch>
                <Route path={buildRoute(match.url, Routes.PROFILE)} component={ProfileContainer} />
                <Route path={buildRoute(match.url, Routes.TOS)} component={TermsPage}/>
            </Switch>
        )
    }
}