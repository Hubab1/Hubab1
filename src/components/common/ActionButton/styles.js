import {
    css
} from 'emotion';
import styled from '@emotion/styled';

export const root = css`
    border-radius: 21.5px !important;
`

export const label = css`
    text-transform: none;
`

export const ButtonContainer = styled.div(props => ({
    marginTop: props.marginTop || 0,
    marginBottom: props.marginBottom || 0
}))
