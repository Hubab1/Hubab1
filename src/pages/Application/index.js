import React, { Component } from 'react';
import { Route, Switch } from 'react-router-dom';

import { ROUTES } from 'constants/constants';
import ApplicationCompletePage from './pages/ApplicationComplete';
import ApplicationApprovedPage from './pages/ApplicationApproved';
import ApplicationDeniedPage from './pages/ApplicationDenied';
import ApplicationCancelledPage from './pages/ApplicationCancelled';

class ApplicationContainer extends Component {
    render() {
        return (
            <Switch>
                <Route path={ROUTES.APP_COMPLETE} component={ApplicationCompletePage} />
                <Route path={ROUTES.APP_APPROVED} component={ApplicationApprovedPage} />
                <Route path={ROUTES.APP_DENIED} component={ApplicationDeniedPage} />
                <Route path={ROUTES.APP_CANCELLED} component={ApplicationCancelledPage} />
            </Switch>
        );
    }
}

export default ApplicationContainer;
