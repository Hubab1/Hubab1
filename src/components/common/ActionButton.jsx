import React from 'react';

import Button from '@material-ui/core/Button';
import { root, label } from 'assets/emotion/ActionButton';

export default class ActionButton extends React.Component {
    render () {
        const { color, disabled, onClick, children } = this.props;
        return (
            <Button
                onClick={onClick}
                classes={{ root, label }}
                variant="contained"
                color={color || 'primary'}
                type="submit"
                disabled={disabled}
                fullWidth
            >
                {children}
            </Button>
        );
    }
}