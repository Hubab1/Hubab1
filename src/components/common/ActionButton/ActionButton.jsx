import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';

import { root, label, ButtonContainer } from './styles';

export default class ActionButton extends React.Component {
    render () {
        const { color, disabled, onClick, children, marginTop, marginBottom, variant } = this.props;
        return (
            <ButtonContainer marginTop={marginTop} marginBottom={marginBottom}>
                <Button
                    onClick={onClick}
                    classes={{ root, label }}
                    variant={variant}
                    color={color}
                    type="submit"
                    disabled={disabled}
                    fullWidth
                >
                    {children}
                </Button>
            </ButtonContainer>
        );
    }
}

ActionButton.propTypes = {
    color: PropTypes.string,
    onClick: PropTypes.func,
    disabled: PropTypes.bool,
    marginTop: PropTypes.string,
    marginBottom: PropTypes.string,
    variant: PropTypes.string,
}

ActionButton.defaultProps = {
    variant: 'contained',
    color: 'primary',
}