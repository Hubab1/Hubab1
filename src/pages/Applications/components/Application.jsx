import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { isEmpty } from 'lodash';
import clsx from 'clsx';
import {
    makeStyles,
    Card,
    CardHeader,
    CardContent,
    Collapse,
    IconButton,
    Typography,
    Divider,
} from '@material-ui/core';
import { ExpandMore as ExpandMoreIcone } from '@material-ui/icons';

import {
    APPLICATION_STATUSES_LABELS,
    APPLICANT_ROLE_LABELS,
    APPLICATION_STATUSES_COLORS,
    APPLICATION_STATUS_DENIED,
} from 'constants/constants';
import ActionButton from 'common-components/ActionButton/ActionButton';
import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import { fetchRenterProfile, selectors } from 'reducers/renter-profile';
import API from 'api/api';
import ErrorIcon from '@material-ui/icons/Error';

const useStyles = makeStyles((theme) => ({
    root: {
        marginBottom: theme.spacing(2),
        '&:last-child': {
            marginBottom: 0,
        },
    },
    expand: {
        transform: 'rotate(0deg)',
        marginLeft: 'auto',
        transition: theme.transitions.create('transform', {
            duration: theme.transitions.duration.shortest,
        }),
    },
    expandOpen: {
        transform: 'rotate(180deg)',
    },
    divider: {
        margin: theme.spacing(0, 2),
    },
    typography: {
        color: '#828796',
        marginBottom: theme.spacing(0.5),

        '& span': {
            color: 'black',
        },
    },
    title: {
        color: '#000',
    },
    applicationId: {
        color: '#828796',
        display: 'block',
        marginTop: theme.spacing(2),
    },
    inviteeNotice: {
        display: 'flex',
        alignItems: 'center',
        color: '#f44336',
        '& span': {
            paddingLeft: '5px',
        },
    },
}));

export function Application({ application = {}, isActive = true, fetchRenterProfile, initialPage, history, invitee }) {
    const { id, status, lease_start_date, lease_term, fees_breakdown, role, unit, community } = application;
    const classes = useStyles();
    const [expanded, setExpanded] = useState(isActive);
    const title = unit?.unit_number ? `${community.display_name}, #${unit.unit_number}` : community.display_name;

    const handleExpandClick = useCallback(() => {
        setExpanded(!expanded);
    }, [expanded]);

    const [appSelected, setAppSelected] = useState(false);
    const [loading, setLoading] = useState(false);

    useEffect(() => {
        if (initialPage && appSelected) {
            setAppSelected(false);
            setLoading(false);
            history.push(initialPage);
        }
    }, [initialPage, appSelected, history]);

    const handleApplicationClick = async (applicationId) => {
        setLoading(true);
        if (invitee) {
            await API.createApplicantRole(invitee.id);
        }
        await fetchRenterProfile(applicationId);
        setAppSelected(true);
    };

    if (isEmpty(application)) {
        return null;
    }

    const showOpenApplicationButton = isActive || application.status === APPLICATION_STATUS_DENIED;

    const inviteeCTA = loading ? 'Starting Application' : 'Start Application';
    const applicationCTA = loading ? 'Loading Application' : 'Go To Application';

    return (
        <Card className={classes.root} elevation={2}>
            <CardHeader
                data-testid="header"
                title={
                    <Typography className={clsx(classes.typography, classes.title)} variant="body1">
                        <b>{title}</b>
                    </Typography>
                }
                subheader={
                    <Typography className={classes.typography} variant="body1">
                        Application Status:{' '}
                        <span style={{ color: APPLICATION_STATUSES_COLORS[status] }}>
                            {APPLICATION_STATUSES_LABELS[status]}
                        </span>
                    </Typography>
                }
                action={
                    <IconButton
                        className={clsx(classes.expand, {
                            [classes.expandOpen]: expanded,
                        })}
                        onClick={handleExpandClick}
                    >
                        <ExpandMoreIcone />
                    </IconButton>
                }
            />
            <Collapse in={expanded}>
                <Divider className={classes.divider} />
                <CardContent data-testid="content">
                    {invitee && (
                        <p className={classes.inviteeNotice}>
                            <ErrorIcon />
                            <span>Youâ€™ve been invited to apply for this unit.</span>
                        </p>
                    )}
                    <Typography className={classes.typography} variant="body1">
                        Move in Date:{' '}
                        <span>{lease_start_date ? moment(lease_start_date).format('MM/DD/YYYY') : '-'}</span>
                    </Typography>
                    <Typography className={classes.typography} variant="body1">
                        Lease Term: <span>{lease_term ? lease_term : '-'}</span>
                    </Typography>
                    <Typography className={classes.typography} variant="body1">
                        Monthly Rent:{' '}
                        <span>
                            {fees_breakdown?.monthly_fees?.total ? `$${fees_breakdown.monthly_fees.total}` : '-'}
                        </span>
                    </Typography>
                    <Typography className={classes.typography} variant="body1">
                        Role: <span>{`${APPLICANT_ROLE_LABELS[role || invitee.role]}`}</span>
                    </Typography>
                    <Typography className={classes.applicationId} variant="caption">
                        Application ID <span>{id}</span>
                    </Typography>
                    {showOpenApplicationButton && (
                        <ActionButton
                            variant="outlined"
                            marginTop={20}
                            marginBottom={10}
                            onClick={() => handleApplicationClick(id)}
                        >
                            {invitee ? inviteeCTA : applicationCTA}
                        </ActionButton>
                    )}
                </CardContent>
            </Collapse>
        </Card>
    );
}

Application.propTypes = {
    application: PropTypes.shape({
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        status: PropTypes.number.isRequired,
        lease_start_date: PropTypes.string,
        lease_term: PropTypes.number,
        fees_breakdown: PropTypes.object,
        role: PropTypes.number.isRequired,
        unit: PropTypes.object,
        community: PropTypes.object.isRequired,
    }),
    isActive: PropTypes.bool,
    invitee: PropTypes.shape({
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        role: PropTypes.number.isRequired,
    }),
    initialPage: PropTypes.string,
    communityId: PropTypes.number,
    fetchRenterProfile: PropTypes.func,
    history: PropTypes.object,
};

const mapStateToProps = (state) => ({
    communityId: state.siteConfig.basename,
    initialPage: selectors.selectDefaultInitialPage(state),
});

const mapDispatchToProps = {
    fetchRenterProfile,
};

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Application));
