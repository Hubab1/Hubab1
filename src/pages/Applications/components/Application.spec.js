import React from 'react';
import { shallow } from 'enzyme';

import {
    ACTIVE_APPLICATION_STATUSES,
    APPLICANT_ROLE_VALUES,
    APPLICATION_STATUSES,
    ROLE_CO_APPLICANT,
    ROUTES,
} from 'constants/constants';
import { Application } from './Application';
import ActionButton from 'common-components/ActionButton/ActionButton';
import API from 'api/api';

describe('Application', () => {
    let application = {};
    beforeEach(() => {
        application = {
            id: 1,
            status: ACTIVE_APPLICATION_STATUSES[0],
            lease_start_date: new Date('01-01-2020').toISOString(),
            lease_term: 12,
            fees_breakdown: {
                monthly_fees: {
                    total: '1.500',
                },
            },
            role: APPLICANT_ROLE_VALUES.ROLE_PRIMARY_APPLICANT,
            unit: {
                unit_number: '12',
            },
            community: {
                display_name: 'Community Alpha',
            },
        };
    });

    it('renders application statuses and colors', () => {
        let wrapper;

        // Black statusses
        wrapper = shallow(
            <Application
                application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_IN_PROGRESS }}
            />
        );
        expect(wrapper.html().includes('<span style="color:#000">In Progress</span>')).toBe(true);

        wrapper = shallow(
            <Application application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_SUBMITTED }} />
        );
        expect(wrapper.html().includes('<span style="color:#000">Submitted</span>')).toBe(true);

        wrapper = shallow(
            <Application application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_COMPLETED }} />
        );
        expect(wrapper.html().includes('<span style="color:#000">Completed</span>')).toBe(true);

        // Green statusses
        wrapper = shallow(
            <Application application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_APPROVED }} />
        );
        expect(wrapper.html().includes('<span style="color:#67C18B">Approved</span>')).toBe(true);

        // Yellow statusses
        wrapper = shallow(
            <Application
                application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_CONDITIONALLY_APPROVED }}
            />
        );
        expect(wrapper.html().includes('<span style="color:#FAC700">Conditionally Approved</span>')).toBe(true);

        // Red statusses
        wrapper = shallow(
            <Application application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_DENIED }} />
        );
        expect(wrapper.html().includes('<span style="color:#FB6D68">Denied</span>')).toBe(true);

        wrapper = shallow(
            <Application application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_CANCELLED }} />
        );
        expect(wrapper.html().includes('<span style="color:#FB6D68">Cancelled</span>')).toBe(true);
    });

    it('renders applicant roles', () => {
        let wrapper;

        wrapper = shallow(
            <Application application={{ ...application, role: APPLICANT_ROLE_VALUES.ROLE_PRIMARY_APPLICANT }} />
        );
        expect(wrapper.html().includes('Main Applicant')).toBe(true);

        wrapper = shallow(
            <Application application={{ ...application, role: APPLICANT_ROLE_VALUES.ROLE_CO_APPLICANT }} />
        );
        expect(wrapper.html().includes('Roommate')).toBe(true);

        wrapper = shallow(<Application application={{ ...application, role: APPLICANT_ROLE_VALUES.ROLE_GUARANTOR }} />);
        expect(wrapper.html().includes('Guarantor')).toBe(true);

        wrapper = shallow(<Application application={{ ...application, role: APPLICANT_ROLE_VALUES.ROLE_OCCUPANT }} />);
        expect(wrapper.html().includes('Occupant')).toBe(true);
    });

    it('renders application - all data present', () => {
        const wrapper = shallow(<Application application={application} isActive={true} />);
        const content = wrapper.find('[data-testid="content"]');
        const header = wrapper.find('[data-testid="header"]');

        expect(header.html().includes('Community Alpha, #12')).toBe(true);
        expect(content.html().includes('Move in Date: <span>01/01/2020</span>')).toBe(true);
        expect(content.html().includes('Lease Term: <span>12</span>')).toBe(true);
        expect(content.html().includes('Monthly Rent: <span>$1.500</span>')).toBe(true);
        expect(content.html().includes('Role: <span>Main Applicant</span>')).toBe(true);
        expect(content.html().includes('Application ID <span>1</span>')).toBe(true);
    });

    it('renders application content - missing data', () => {
        application = {
            ...application,
            unit: undefined,
            lease_start_date: undefined,
            lease_term: undefined,
            fees_breakdown: undefined,
        };

        const wrapper = shallow(<Application application={application} isActive={true} />);
        const content = wrapper.find('[data-testid="content"]');
        const header = wrapper.find('[data-testid="header"]');

        expect(header.html().includes('Community Alpha')).toBe(true);
        expect(content.html().includes('Move in Date: <span>-</span>')).toBe(true);
        expect(content.html().includes('Lease Term: <span>-</span>')).toBe(true);
        expect(content.html().includes('Monthly Rent: <span>-</span>')).toBe(true);
        expect(content.html().includes('Role: <span>Main Applicant</span>')).toBe(true);
        expect(content.html().includes('Application ID <span>1</span>')).toBe(true);
    });

    it('renders "open application" button if active app or past denied app', () => {
        let wrapper;

        wrapper = shallow(
            <Application
                application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_IN_PROGRESS }}
            />
        );
        expect(wrapper.find(ActionButton).length).toBe(1);

        wrapper = shallow(
            <Application application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_COMPLETED }} />
        );
        expect(wrapper.find(ActionButton).length).toBe(1);

        wrapper = shallow(
            <Application
                isActive={false}
                application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_DENIED }}
            />
        );
        expect(wrapper.find(ActionButton).length).toBe(1);

        wrapper = shallow(
            <Application
                isActive={false}
                application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_COMPLETED }}
            />
        );
        expect(wrapper.find(ActionButton).length).toBe(0);

        wrapper = shallow(
            <Application
                isActive={false}
                application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_CANCELLED }}
            />
        );
        expect(wrapper.find(ActionButton).length).toBe(0);
    });

    it('fetches an application when clicking on the button', async () => {
        const fetchRenterProfile = jest.fn();
        const wrapper = shallow(
            <Application
                application={{ ...application, status: APPLICATION_STATUSES.APPLICATION_STATUS_IN_PROGRESS }}
                fetchRenterProfile={fetchRenterProfile}
                initnialPage={ROUTES.FEES_AND_DEPOSITS}
            />
        );
        await wrapper.find(ActionButton).simulate('click');
        expect(fetchRenterProfile).toBeCalledWith(application.id);
    });

    describe('invitation', () => {
        const invitee = {
            id: 123,
            role: 30,
        };

        const fetchRenterProfile = jest.fn().mockReturnValue(Promise.resolve());
        let wrapper;

        beforeEach(() => {
            wrapper = shallow(
                <Application
                    application={{ ...application }}
                    invitee={invitee}
                    fetchRenterProfile={fetchRenterProfile}
                />
            );
        });

        afterEach(() => {
            jest.restoreAllMocks();
        });

        it('renders invitations correctly', async () => {
            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('handles "Start Application" button correctly', async () => {
            jest.spyOn(API, 'createApplicantRole').mockReturnValue({});
            await wrapper.find(ActionButton).simulate('click');
            expect(API.createApplicantRole).toBeCalledWith(invitee.id);
            expect(fetchRenterProfile).toBeCalledWith(application.id);
        });
    });
});
