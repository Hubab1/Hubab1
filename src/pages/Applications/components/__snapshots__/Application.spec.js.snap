// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application invitation renders invitations correctly 1`] = `
<ForwardRef(WithStyles)
  className="makeStyles-root-1"
  elevation={2}
>
  <ForwardRef(WithStyles)
    action={
      <ForwardRef(WithStyles)
        className="makeStyles-expand-2 makeStyles-expandOpen-3"
        onClick={[Function]}
      >
        <UNDEFINED />
      </ForwardRef(WithStyles)>
    }
    data-testid="header"
    subheader={
      <ForwardRef(WithStyles)
        className="makeStyles-typography-5"
        variant="body1"
      >
        Application Status:
         
        <span
          style={
            Object {
              "color": "#000",
            }
          }
        >
          In Progress
        </span>
      </ForwardRef(WithStyles)>
    }
    title={
      <ForwardRef(WithStyles)
        className="makeStyles-typography-5 makeStyles-title-6"
        variant="body1"
      >
        <b>
          Community Alpha, #12
        </b>
      </ForwardRef(WithStyles)>
    }
  />
  <ForwardRef(WithStyles)
    in={true}
  >
    <ForwardRef(WithStyles)
      className="makeStyles-divider-4"
    />
    <ForwardRef(WithStyles)
      data-testid="content"
    >
      <p
        className="makeStyles-inviteeNotice-8"
      >
        <UNDEFINED />
        <span>
          Youâ€™ve been invited to apply for this unit.
        </span>
      </p>
      <ForwardRef(WithStyles)
        className="makeStyles-typography-5"
        variant="body1"
      >
        Move in Date:
         
        <span>
          01/01/2020
        </span>
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)
        className="makeStyles-typography-5"
        variant="body1"
      >
        Lease Term: 
        <span>
          12
        </span>
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)
        className="makeStyles-typography-5"
        variant="body1"
      >
        Monthly Rent:
         
        <span>
          $1.500
        </span>
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)
        className="makeStyles-typography-5"
        variant="body1"
      >
        Role: 
        <span>
          Main Applicant
        </span>
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)
        className="makeStyles-applicationId-7"
        variant="caption"
      >
        Application ID 
        <span>
          1
        </span>
      </ForwardRef(WithStyles)>
      <ActionButton
        color="primary"
        marginBottom={10}
        marginTop={20}
        onClick={[Function]}
        successGreen={false}
        variant="outlined"
      >
        Start Application
      </ActionButton>
    </ForwardRef(WithStyles)>
  </ForwardRef(WithStyles)>
</ForwardRef(WithStyles)>
`;
